content/6_Advanced/Wavelet.mdx

```diff
@@ -32,9 +30,234 @@ Like a segment tree on values rather than indices.
+
+Let's say our array is: $[3,5,3,1,2,2,3,4,5,5]$
+Each node has an array representing the indices of every number between l and r
+
+![Wavelet Tree Visualization](./assets/diagram.png)
+</Spoiler>
+
 ### Solution - Range K-th Smallest
+Before we solve this problem, let's consider a simpler version where we are asked, given a range, to count the number of occurrences of value $x$.
+
+## Solving the first type of query
+**Given a range $l$, $r$, count the number of occurrences of value x.**
+
+To calculate the number of occurrences from $𝑙$ to $𝑟$, we can use the following
+formula:
+
+$$
+\begin{aligned}
+\texttt{occurrences}(l, r) = \texttt{occurrences}(r) - \texttt{occurrences}(l)
+\end{aligned}
+$$
+
+This reduces the problem to counting the number of occurrences in a prefix.
+
+One way to solve the problem is to go to the leaf node
+and perform a binary search for the number of indices less than $𝑟$
+However, let's explore a different approach that can also be extended to the
+second type of query.
+
+Instead of binary searching on the leaf, we update $𝑟$ as we recurse down the
+tree.
+If we can determine the position (index) of $r$ in the left and right children of a node, we can recurse down the tree and determine its position in the leaf node.
+
+To find the position of $𝑟$ in a node's left and right children, we need to
+determine how many indices are smaller than the middle value (mid) and precede
+$𝑟$.
+This can be done using a prefix sum.
+
+Let's define:
+- $c[i]$ = as 1 if $index[i]$ is smaller than mid otherwise 0
+- $prefix_b[i]$ as prefix sum of $c[i]$
+
+Formally
+
+$$
+c[i] = \begin{cases}
+    1, & \text{if } \text{index}[i] < \text{mid} \\
+    0, & \text{otherwise}
+\end{cases}
+$$
+$$
+\text{prefixB}[i] = \text{prefixB}[i - 1] + c[i]
+$$
+
+
+To update $r$ as we recurse down, we do the following:
+- To know the value of $r$ if we recurse left, we use prefix_b[r]
+- If we recurse right, we use $r$ - prefix_b[r]
+
+Now let's try to solve our main problem.
+## Solving the second type of query
+**Given a range $l$, $r$ find the k smallest element**
+
+We will determine whether the answer for a given node is in the left or the
+right segment.
+We can calculate how many times the elements within the segments' ranges appear
+in our range $(l, r)$ using our first type of query.
+Note that this also works for non-leaf nodes using the following formula:
+
+$$
+\texttt{occurrences}(l, r) = r - l
+$$
+<Info title="Similar">
+This is similar to counting how many times a value appears up to index $r$ in our previous query. We did this by using the new $r$ value at the leaf node. But now, we consider the difference between the updated $r$ and $l$
+</Info>
+
+Therefore, the occurrences of the left node is
+
+$$
+\texttt{left\_occurrences} = prefix_b[r] - prefix_b[l]
+$$
+
+<Info title="Left occurrences">
+  Note that $\texttt{left\_occurrences}$ is the number of indices between l and r whose value is less than mid
+
+</Info>
```

Comment:
This doesn't need to be an info block, you can just include it in the main content.